fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java

```diff
@@ -48,32 +64,139 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
+                                                   OptimizerContext context) {
+        if (apply.isNeedCheckMaxRows()) {
+            return transformCorrelateWithCheckOneRows(input, apply, context);
+        } else {
+            return transformCorrelateWithoutCheckOneRows(input, apply, context);
+        }
+    }
 
-            joinOptExpression = new OptExpression(
-                    new LogicalJoinOperator(JoinOperator.LEFT_OUTER_JOIN,
-                            Utils.compoundAnd(apply.getCorrelationConjuncts(), apply.getPredicate())));
+    /*
+     * e.g.
+     *   SELECT * FROM t0
+     *   WHERE t0.v2 > (
+     *       SELECT t1.v2 FROM t1
+     *       WHERE t0.v1 = t1.v1
+     *   );
+     *
+     * Transfer to:
+     *   SELECT t0.*
+     *   FROM t0 LEFT OUTER JOIN
+     *   (
+     *       SELECT t1.v1, count(1) as countRows, any_value(t1.v2) as anyValue
+     *       FROM t1
+     *       GROUP BY t1.v1
+     *   ) as t1a
+     *   ON t0.v1 = t1a.v1
+     *   WHERE t0.v2 > t1a.anyValue;
+     */
+    private List<OptExpression> transformCorrelateWithCheckOneRows(OptExpression input, LogicalApplyOperator apply,
+                                                                   OptimizerContext context) {
+        // t0.v1 = t1.v1
+        ScalarOperator correlationPredicate = apply.getCorrelationConjuncts();
+        Pair<List<ScalarOperator>, Map<ColumnRefOperator, ScalarOperator>> correlationPredicatePair =
+                SubqueryUtils.rewritePredicateAndExtractColumnRefs(Utils.extractConjuncts(correlationPredicate),
+                        apply.getCorrelationColumnRefs(), context);
+        // t1.v1
+        ColumnRefSet correlationPredicateInnerRefs = new ColumnRefSet(correlationPredicatePair.second.keySet());
+
+        /*
+         * Step1: build agg
+         *      output: count(1) as countRows, any_value(t1.v2) as anyValue
+         *      groupBy: t1.v1
+         */
+        List<ColumnRefOperator> countAggregateGroupBys = Lists.newArrayList();
 
-            joinOptExpression.getInputs().addAll(input.getInputs());
+        ColumnRefFactory factory = context.getColumnRefFactory();
+        for (int columnId : correlationPredicateInnerRefs.getColumnIds()) {
+            ColumnRefOperator ref = factory.getColumnRef(columnId);
+            countAggregateGroupBys.add(ref);
+        }
 
-            Map<ColumnRefOperator, ScalarOperator> output = Maps.newHashMap();
-            output.put(apply.getOutput(), apply.getSubqueryOperator());
+        // count aggregate
+        Map<ColumnRefOperator, CallOperator> aggregates = Maps.newHashMap();
+        CallOperator countRowsCallOp = SubqueryUtils.createCountRowsOperator();
+        ColumnRefOperator countRows =
+                factory.create("countRows", countRowsCallOp.getType(), countRowsCallOp.isNullable());
+        aggregates.put(countRows, countRowsCallOp);
+
+        // any_value aggregate
+        ScalarOperator subqueryOperator = apply.getSubqueryOperator();
```

Comment:
Would better put all create concrete aggregate function call to a util class, then we could reuse some codes.

Reply:
done